File: packages/core/src/storage/document-source.ts
import { Effect, Schema } from "effect";
import { ValidationError } from "../errors/crud-errors.js";
import type { MigrationError } from "../errors/migration-errors.js";
import {
	DuplicatePhysicalFileError,
	DuplicateRecordError,
	InvalidDocumentSourceError,
	type SourceError,
	UnknownCollectionError,
} from "../errors/source-errors.js";
import {
	SerializationError,
	StorageError,
	type UnsupportedFormatError,
} from "../errors/storage-errors.js";
import { runMigrations } from "../migrations/migration-runner.js";
import { SerializerRegistry } from "../serializers/serializer-service.js";
import type { DerivedIdConfig } from "../types/database-config-types.js";
import type { HasId } from "./derived-id.js";
import {
	assertNoPhysicalDerivedId,
	hydrateDerivedId,
	requireHydratablePayload,
	stripDerivedIdField,
} from "./derived-id.js";
import { type OriginIndex, setOrigin } from "./origin-index.js";
import {
	matchesDocumentSourcePattern,
	type NormalizedDocumentSourceConfig,
	type NormalizedSourceConfig,
} from "./source-config.js";
import { StorageAdapter } from "./storage-service.js";

export interface LoadedDocumentSources {
	readonly collections: Record<string, ReadonlyMap<string, HasId>>;
	readonly origins: OriginIndex;
	readonly documents: ReadonlyArray<LoadedDocument>;
}

export interface SavedDocumentSource {
	readonly origins: OriginIndex;
	readonly documents: ReadonlyArray<LoadedDocument>;
}

export interface SaveDocumentSourceInput {
	readonly config: NormalizedSourceConfig;
	readonly sourceId: string;
	readonly collections: Record<string, ReadonlyMap<string, HasId>>;
	readonly origins: OriginIndex;
	readonly documents: ReadonlyArray<LoadedDocument>;
}

export interface LoadedDocument {
	readonly sourceId: string;
	readonly path: string;
	readonly data: Record<string, unknown>;
}

type CollectionLoadConfig = {
	readonly name: string;
	readonly schema: Schema.Codec<HasId, unknown, never, never>;
	readonly id?: DerivedIdConfig;
	readonly version?: number;
	readonly migrations?: ReadonlyArray<
		import("../migrations/migration-types.js").Migration
	>;
};

const isRecord = (value: unknown): value is Record<string, unknown> =>
	typeof value === "object" && value !== null && !Array.isArray(value);

const emptyCollectionMaps = (
	collectionNames: ReadonlyArray<string>,
): Record<string, Map<string, HasId>> => {
	const collections: Record<string, Map<string, HasId>> = {};
	for (const collectionName of collectionNames) {
		collections[collectionName] = new Map<string, HasId>();
	}
	return collections;
};

export const loadDocumentSources = (
	config: NormalizedSourceConfig,
): Effect.Effect<
	LoadedDocumentSources,
	| StorageError
	| SerializationError
	| UnsupportedFormatError
	| ValidationError
	| MigrationError
	| SourceError,
	StorageAdapter | SerializerRegistry
> =>
	Effect.gen(function* () {
		const collectionConfigs = new Map<string, CollectionLoadConfig>();
		for (const collectionName of config.collections) {
			const collectionConfig = config.collectionConfigs[collectionName];
			collectionConfigs.set(collectionName, {
				name: collectionName,
				schema: collectionConfig.schema as Schema.Codec<
					HasId,
					unknown,
					never,
					never
				>,
				...(collectionConfig.id !== undefined
					? { id: collectionConfig.id }
					: {}),
				...(collectionConfig.version !== undefined
					? { version: collectionConfig.version }
					: {}),
				...(collectionConfig.migrations !== undefined
					? { migrations: collectionConfig.migrations }
					: {}),
			});
		}

		const collections = emptyCollectionMaps(config.collections);
		const origins = new Map<string, import("./origin-index.js").RecordOrigin>();
		const documents: LoadedDocument[] = [];

		for (const source of config.sources) {
			// documentGraph sources are loaded separately by loadDocumentGraphSources;
			// the writable documents loader only handles `documents` sources.
			if (source.kind !== "documents") continue;
			yield* loadDocumentSource(
				source,
				collectionConfigs,
				collections,
				origins,
				documents,
			);
		}

		return {
			collections,
			origins,
			documents,
		};
	});

export const saveDocumentSource = (
	input: SaveDocumentSourceInput,
): Effect.Effect<
	SavedDocumentSource,
	| StorageError
	| SerializationError
	| UnsupportedFormatError
	| ValidationError
	| SourceError,
	StorageAdapter | SerializerRegistry
> =>
	Effect.gen(function* () {
		const source = input.config.sources.find(
			(candidate) => candidate.id === input.sourceId,
		);
		if (source === undefined) {
			return yield* Effect.fail(
				new InvalidDocumentSourceError({
					sourceId: input.sourceId,
					path: "",
					message: `Unknown document source '${input.sourceId}'`,
				}),
			);
		}
		if (source.kind !== "documents") {
			// documentGraph (and any future read-only source) has no write path.
			return yield* Effect.fail(
				new InvalidDocumentSourceError({
					sourceId: input.sourceId,
					path: "",
					message: `Source '${input.sourceId}' is read-only and cannot be saved`,
				}),
			);
		}

		const collectionConfigs = new Map<string, CollectionLoadConfig>();
		for (const collectionName of source.collections) {
			const collectionConfig = input.config.collectionConfigs[collectionName];
			collectionConfigs.set(collectionName, {
				name: collectionName,
				schema: collectionConfig.schema as Schema.Codec<
					HasId,
					unknown,
					never,
					never
				>,
				...(collectionConfig.id !== undefined
					? { id: collectionConfig.id }
					: {}),
				...(collectionConfig.version !== undefined
					? { version: collectionConfig.version }
					: {}),
				...(collectionConfig.migrations !== undefined
					? { migrations: collectionConfig.migrations }
					: {}),
			});
		}

		const projectedByPath = new Map<string, Record<string, unknown>>();
		const sourceDocuments = input.documents.filter(
			(document) => document.sourceId === source.id,
		);
		for (const document of sourceDocuments) {
			projectedByPath.set(
				document.path,
				cloneDocumentForSource(source, document, collectionConfigs),
			);
		}
		const newOrigins = new Map(input.origins);
		for (const [key, origin] of input.origins) {
			if (origin.sourceId !== source.id) continue;
			if (!source.collections.includes(origin.collection)) continue;
			const currentCollection = input.collections[origin.collection];
			if (
				currentCollection === undefined ||
				!currentCollection.has(origin.id)
			) {
				newOrigins.delete(key);
			}
		}

		for (const collectionName of source.collections) {
			const collectionConfig = collectionConfigs.get(collectionName);
			const currentCollection = input.collections[collectionName];
			if (collectionConfig === undefined || currentCollection === undefined) {
				continue;
			}

			for (const [id, entity] of currentCollection) {
				const existingOrigin = input.origins.get(
					`${collectionName}\u0000${id}`,
				);
				if (
					existingOrigin !== undefined &&
					existingOrigin.sourceId !== source.id
				) {
					continue;
				}
				const path = existingOrigin?.path ?? source.outbox;
				const targetDocument = getOrCreateDocument(projectedByPath, path);
				const section = getOrCreateSection(targetDocument, collectionName);
				stampSectionVersion(section, collectionConfig);
				const encoded = yield* encodeDocumentSourceEntity(
					source,
					path,
					collectionConfig,
					id,
					entity,
				);
				section[id] = encoded;
				setOrigin(newOrigins, {
					sourceId: source.id,
					path,
					collection: collectionName,
					id,
				});
			}
		}

		const serializer = yield* SerializerRegistry;
		const storage = yield* StorageAdapter;
		const writes: Array<{ readonly path: string; readonly data: string }> = [];
		for (const [path, document] of projectedByPath) {
			const data = yield* serializer.serialize(document, source.format);
			writes.push({ path, data });
		}

		for (const write of writes) {
			yield* storage.ensureDir(write.path);
			yield* storage.write(write.path, write.data);
		}

		const mergedDocuments = [
			...input.documents.filter((document) => document.sourceId !== source.id),
			...Array.from(projectedByPath.entries()).map(([path, data]) => ({
				sourceId: source.id,
				path,
				data,
			})),
		].sort((left, right) =>
			left.sourceId === right.sourceId
				? left.path.localeCompare(right.path)
				: left.sourceId.localeCompare(right.sourceId),
		);

		return {
			origins: newOrigins,
			documents: mergedDocuments,
		};
	});

const cloneDocumentForSource = (
	source: NormalizedDocumentSourceConfig,
	document: LoadedDocument,
	collectionConfigs: ReadonlyMap<string, CollectionLoadConfig>,
): Record<string, unknown> => {
	const cloned: Record<string, unknown> = { ...document.data };
	for (const collectionName of source.collections) {
		const section = cloned[collectionName];
		if (!isRecord(section)) continue;
		const nextSection: Record<string, unknown> = {};
		const collectionConfig = collectionConfigs.get(collectionName);
		stampSectionVersion(nextSection, collectionConfig);
		if (
			collectionConfig?.version === undefined &&
			section._version !== undefined
		) {
			nextSection._version = section._version;
		}
		cloned[collectionName] = nextSection;
	}
	return cloned;
};

const getOrCreateDocument = (
	documents: Map<string, Record<string, unknown>>,
	path: string,
): Record<string, unknown> => {
	const existing = documents.get(path);
	if (existing !== undefined) return existing;
	const created: Record<string, unknown> = {};
	documents.set(path, created);
	return created;
};

const getOrCreateSection = (
	document: Record<string, unknown>,
	collectionName: string,
): Record<string, unknown> => {
	const existing = document[collectionName];
	if (isRecord(existing)) return existing;
	const created: Record<string, unknown> = {};
	document[collectionName] = created;
	return created;
};

const stampSectionVersion = (
	section: Record<string, unknown>,
	collectionConfig: CollectionLoadConfig | undefined,
): void => {
	if (collectionConfig?.version !== undefined) {
		section._version = collectionConfig.version;
	}
};

const encodeDocumentSourceEntity = (
	source: NormalizedDocumentSourceConfig,
	path: string,
	collectionConfig: CollectionLoadConfig,
	id: string,
	entity: HasId,
): Effect.Effect<unknown, ValidationError, never> => {
	const encode = Schema.encodeEffect(collectionConfig.schema);
	const valueToEncode = stripDerivedIdField(entity, collectionConfig.id);
	return encode(valueToEncode as HasId).pipe(
		Effect.mapError(
			(parseError) =>
				new ValidationError({
					message: `Failed to encode entity '${id}' in collection '${collectionConfig.name}' for document source '${source.id}' at '${path}': ${parseError.message}`,
					issues: [
						{
							field: `${collectionConfig.name}.${id}`,
							message: parseError.message,
						},
					],
				}),
		),
	);
};
const loadDocumentSource = (
	source: NormalizedDocumentSourceConfig,
	collectionConfigs: ReadonlyMap<string, CollectionLoadConfig>,
	collections: Record<string, Map<string, HasId>>,
	origins: Map<string, import("./origin-index.js").RecordOrigin>,
	documents: Array<LoadedDocument>,
): Effect.Effect<
	void,
	| StorageError
	| SerializationError
	| UnsupportedFormatError
	| ValidationError
	| MigrationError
	| SourceError,
	StorageAdapter | SerializerRegistry
> =>
	Effect.gen(function* () {
		const storage = yield* StorageAdapter;
		const rootExists = yield* storage.exists(source.root);
		if (!rootExists && !source.optional) {
			return yield* Effect.fail(
				new StorageError({
					path: source.root,
					operation: "list",
					message: `Document source root '${source.root}' does not exist`,
				}),
			);
		}
		if (!rootExists && source.optional) return;

		const discovered = yield* storage.listRecursive(source.root);
		const seenPaths = new Set<string>();
		const matchingFiles: string[] = [];
		for (const path of discovered) {
			if (!matchesDocumentSourcePattern(source, path)) continue;
			if (seenPaths.has(path)) {
				return yield* Effect.fail(
					new DuplicatePhysicalFileError({
						sourceId: source.id,
						path,
						message: `Document source '${source.id}' discovered '${path}' more than once`,
					}),
				);
			}
			seenPaths.add(path);
			matchingFiles.push(path);
		}
		matchingFiles.sort();

		for (const path of matchingFiles) {
			const raw = yield* storage.read(path);
			const trimmed = raw.trim();
			if (trimmed.length === 0) {
				documents.push({ sourceId: source.id, path, data: {} });
				continue;
			}

			const parsed = yield* deserializeDocument(path, source.format, raw);
			if (parsed === null || parsed === undefined) {
				documents.push({ sourceId: source.id, path, data: {} });
				continue;
			}
			if (!isRecord(parsed)) {
				return yield* Effect.fail(
					new InvalidDocumentSourceError({
						sourceId: source.id,
						path,
						message: `Document source '${source.id}' file '${path}' must contain a top-level object`,
					}),
				);
			}

			documents.push({ sourceId: source.id, path, data: parsed });
			yield* loadDocumentFile(
				source,
				path,
				parsed,
				collectionConfigs,
				collections,
				origins,
			);
		}
	});

const deserializeDocument = (
	path: string,
	format: string,
	raw: string,
): Effect.Effect<
	unknown,
	SerializationError | UnsupportedFormatError,
	SerializerRegistry
> =>
	Effect.gen(function* () {
		const serializer = yield* SerializerRegistry;
		return yield* serializer.deserialize(raw, format).pipe(
			Effect.mapError((error) => {
				if (error._tag !== "SerializationError") return error;
				return new SerializationError({
					format: error.format,
					message: `${error.message} in '${path}'`,
					cause: error.cause,
				});
			}),
		);
	});

const loadDocumentFile = (
	source: NormalizedDocumentSourceConfig,
	path: string,
	parsed: Record<string, unknown>,
	collectionConfigs: ReadonlyMap<string, CollectionLoadConfig>,
	collections: Record<string, Map<string, HasId>>,
	origins: Map<string, import("./origin-index.js").RecordOrigin>,
): Effect.Effect<void, ValidationError | MigrationError | SourceError, never> =>
	Effect.gen(function* () {
		const allowedCollections = new Set(source.collections);
		for (const key of Object.keys(parsed)) {
			if (!allowedCollections.has(key)) {
				if (source.unknownCollections === "preserve") continue;
				return yield* Effect.fail(
					new UnknownCollectionError({
						sourceId: source.id,
						path,
						collection: key,
						message: `Document source '${source.id}' file '${path}' contains unknown collection '${key}'`,
					}),
				);
			}
		}

		for (const collectionName of source.collections) {
			const section = parsed[collectionName];
			if (section === undefined) continue;
			const collectionConfig = collectionConfigs.get(collectionName);
			if (collectionConfig === undefined) continue;
			if (!isRecord(section)) {
				return yield* Effect.fail(
					new InvalidDocumentSourceError({
						sourceId: source.id,
						path,
						collection: collectionName,
						message: `Collection '${collectionName}' in '${path}' must be an object keyed by record id`,
					}),
				);
			}

			yield* loadCollectionSection(
				source,
				path,
				collectionConfig,
				section,
				collections,
				origins,
			);
		}
	});

const loadCollectionSection = (
	source: NormalizedDocumentSourceConfig,
	path: string,
	collectionConfig: CollectionLoadConfig,
	section: Record<string, unknown>,
	collections: Record<string, Map<string, HasId>>,
	origins: Map<string, import("./origin-index.js").RecordOrigin>,
): Effect.Effect<
	void,
	| ValidationError
	| MigrationError
	| DuplicateRecordError
	| InvalidDocumentSourceError,
	never
> =>
	Effect.gen(function* () {
		const fileVersion =
			typeof section._version === "number" ? section._version : 0;
		const entityMap: Record<string, unknown> = {};
		for (const [id, value] of Object.entries(section)) {
			if (id !== "_version") entityMap[id] = value;
		}

		let dataToLoad = entityMap;
		let migrationsRan = false;
		if (collectionConfig.version !== undefined) {
			if (fileVersion > collectionConfig.version) {
				return yield* Effect.fail(
					new InvalidDocumentSourceError({
						sourceId: source.id,
						path,
						collection: collectionConfig.name,
						message: `File version ${fileVersion} for collection '${collectionConfig.name}' is ahead of config version ${collectionConfig.version}`,
					}),
				);
			}
			if (
				fileVersion < collectionConfig.version &&
				collectionConfig.migrations !== undefined &&
				collectionConfig.migrations.length > 0
			) {
				dataToLoad = yield* runMigrations(
					entityMap,
					fileVersion,
					collectionConfig.version,
					collectionConfig.migrations,
					collectionConfig.name,
				);
				migrationsRan = true;
			}
		}

		const decode = Schema.decodeUnknownEffect(collectionConfig.schema);
		for (const [id, value] of Object.entries(dataToLoad)) {
			if (value === null || value === undefined || !isRecord(value)) {
				return yield* Effect.fail(
					new InvalidDocumentSourceError({
						sourceId: source.id,
						path,
						collection: collectionConfig.name,
						id,
						message: `Record '${collectionConfig.name}.${id}' in '${path}' must be an object`,
					}),
				);
			}

			const physicalIdError = assertNoPhysicalDerivedId(
				id,
				value,
				collectionConfig.id,
				`${path}.${collectionConfig.name}`,
			);
			if (physicalIdError !== undefined) {
				return yield* Effect.fail(physicalIdError);
			}

			const decoded = yield* decode(value).pipe(
				Effect.mapError((parseError) =>
					migrationsRan
						? new ValidationError({
								message: `Post-migration validation failed for entity '${id}' in collection '${collectionConfig.name}' from '${path}': ${parseError.message}`,
								issues: [
									{
										field: `${collectionConfig.name}.${id}`,
										message: parseError.message,
									},
								],
							})
						: new ValidationError({
								message: `Failed to decode entity '${id}' in collection '${collectionConfig.name}' from '${path}': ${parseError.message}`,
								issues: [
									{
										field: `${collectionConfig.name}.${id}`,
										message: parseError.message,
									},
								],
							}),
				),
			);
			const hydratableError = requireHydratablePayload(
				id,
				decoded,
				collectionConfig.id,
				`${path}.${collectionConfig.name}`,
			);
			if (hydratableError !== undefined) {
				return yield* Effect.fail(hydratableError);
			}

			const hydrated = hydrateDerivedId<HasId>(
				id,
				decoded,
				collectionConfig.id,
			);
			const target = collections[collectionConfig.name];
			const existingOrigin = origins.get(
				`${collectionConfig.name}\u0000${hydrated.id}`,
			);
			const origin = {
				sourceId: source.id,
				path,
				collection: collectionConfig.name,
				id: hydrated.id,
			};
			if (existingOrigin !== undefined) {
				return yield* Effect.fail(
					new DuplicateRecordError({
						collection: collectionConfig.name,
						id: hydrated.id,
						first: existingOrigin,
						duplicate: origin,
						message: `Duplicate record '${collectionConfig.name}.${hydrated.id}' in '${existingOrigin.path}' and '${path}'`,
					}),
				);
			}
			target.set(hydrated.id, hydrated);
			setOrigin(origins, origin);
		}

		if (collectionConfig.version !== undefined) {
			for (const key of Object.keys(section)) {
				delete section[key];
			}
			section._version = collectionConfig.version;
			for (const [id, value] of Object.entries(dataToLoad)) {
				section[id] = value;
			}
		}
	});

File: crates/proseql-storage/src/document_source.rs
use indexmap::IndexMap;
use proseql_engine::errors::{
    DuplicatePhysicalFileError, DuplicateRecordError, EngineError, InvalidDocumentSourceError,
    SourceConfigError, SourceRecordOrigin, UnknownCollectionError,
};
use proseql_formats::FormatRegistry;
use serde_json::{Map, Value};

use crate::host::StorageHost;
use crate::persistence::{
    decode_entity_map, encode_value, is_version_sidecar_path, run_migrations,
    strip_derived_id_field, MigrationHost, ValidationMode,
};
use crate::source_config::{
    matches_document_source_pattern, NormalizedDatabaseSourceConfig,
    NormalizedDocumentSourceConfig, NormalizedSourceConfig, UnknownCollectionPolicy,
};

#[derive(Debug, Clone, PartialEq)]
pub struct LoadedDocument {
    pub source_id: String,
    pub path: String,
    pub data: Map<String, Value>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LoadedDocumentSources {
    pub collections: IndexMap<String, IndexMap<String, Value>>,
    pub origins: IndexMap<String, SourceRecordOrigin>,
    pub documents: Vec<LoadedDocument>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SavedDocumentSource {
    pub origins: IndexMap<String, SourceRecordOrigin>,
    pub documents: Vec<LoadedDocument>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SaveDocumentSourceInput {
    pub config: NormalizedSourceConfig,
    pub source_id: String,
    pub collections: IndexMap<String, IndexMap<String, Value>>,
    pub origins: IndexMap<String, SourceRecordOrigin>,
    pub documents: Vec<LoadedDocument>,
}

fn origin_key(collection: &str, id: &str) -> String {
    format!("{collection}\u{0}{id}")
}

fn collection_maps(names: &[String]) -> IndexMap<String, IndexMap<String, Value>> {
    names
        .iter()
        .map(|name| (name.clone(), IndexMap::new()))
        .collect()
}

fn duplicate_record_error(
    collection: &str,
    id: &str,
    first: SourceRecordOrigin,
    duplicate: SourceRecordOrigin,
) -> EngineError {
    EngineError::DuplicateRecord(Box::new(DuplicateRecordError {
        collection: collection.to_owned(),
        id: id.to_owned(),
        first,
        duplicate,
        message: format!("Duplicate record '{collection}/{id}' across document sources"),
    }))
}

fn source_config_error(
    message: impl Into<String>,
    source_id: &str,
    collection: Option<&str>,
    path: Option<&str>,
) -> EngineError {
    EngineError::SourceConfig(Box::new(SourceConfigError {
        message: message.into(),
        source_id: Some(source_id.to_owned()),
        collection: collection.map(str::to_owned),
        path: path.map(str::to_owned),
    }))
}

fn collection_config<'a>(
    config: &'a NormalizedSourceConfig,
    source_id: &str,
    collection: &str,
    path: Option<&str>,
) -> Result<&'a crate::persistence::CollectionStorageConfig, EngineError> {
    config.collection_configs.get(collection).ok_or_else(|| {
        source_config_error(
            format!(
                "Source '{}' references collection '{}' without a declared collection config",
                source_id, collection
            ),
            source_id,
            Some(collection),
            path,
        )
    })
}

fn validate_document_source_config(
    config: &NormalizedSourceConfig,
    source: &NormalizedDocumentSourceConfig,
) -> Result<(), EngineError> {
    for collection in &source.collections {
        let _ = collection_config(config, &source.id, collection, Some(&source.root))?;
        if !config.collections.contains(collection) {
            return Err(source_config_error(
                format!(
                    "Source '{}' references collection '{}' outside the normalized collection list",
                    source.id, collection
                ),
                &source.id,
                Some(collection),
                Some(&source.root),
            ));
        }
    }
    Ok(())
}

fn clone_document_for_source(
    document: &LoadedDocument,
    source: &NormalizedDocumentSourceConfig,
    config: &NormalizedSourceConfig,
) -> Map<String, Value> {
    let mut cloned = document.data.clone();
    for collection_name in &source.collections {
        let Some(Value::Object(section)) = cloned.get_mut(collection_name) else {
            continue;
        };
        let mut next = Map::new();
        if let Some(version) = config
            .collection_configs
            .get(collection_name)
            .and_then(|cfg| cfg.version)
        {
            next.insert("_version".to_owned(), Value::Number(version.into()));
        } else if let Some(version) = section.get("_version") {
            next.insert("_version".to_owned(), version.clone());
        }
        *section = next;
    }
    cloned
}

pub fn load_document_sources(
    host: &dyn StorageHost,
    formats: &FormatRegistry,
    config: &NormalizedSourceConfig,
    migration_host: Option<&dyn MigrationHost>,
) -> Result<LoadedDocumentSources, EngineError> {
    let mut collections = collection_maps(&config.collections);
    let mut origins = IndexMap::new();
    let mut documents = Vec::new();

    for source in &config.sources {
        let NormalizedDatabaseSourceConfig::Documents(source) = source else {
            continue;
        };
        validate_document_source_config(config, source)?;
        if !host.exists(&source.root)? {
            if source.optional {
                continue;
            }
            return Err(EngineError::Storage(Box::new(
                proseql_engine::errors::StorageError {
                    path: source.root.clone(),
                    operation: proseql_engine::errors::StorageOperation::List,
                    message: format!("Source root '{}' does not exist", source.root),
                    cause: None,
                },
            )));
        }
        let mut seen_paths = std::collections::HashSet::new();
        let mut matched = host
            .list_recursive(&source.root)?
            .into_iter()
            .filter(|path| {
                matches_document_source_pattern(source, path)
                    && !is_version_sidecar_path(path, &source.format)
            })
            .collect::<Vec<_>>();
        matched.sort();
        for path in matched {
            if !seen_paths.insert(path.clone()) {
                return Err(EngineError::DuplicatePhysicalFile(Box::new(
                    DuplicatePhysicalFileError {
                        source_id: source.id.clone(),
                        path: path.clone(),
                        message: format!(
                            "Document source '{}' discovered '{}' more than once",
                            source.id, path
                        ),
                    },
                )));
            }
            let raw = host.read(&path)?;
            if raw.trim().is_empty() {
                documents.push(LoadedDocument {
                    source_id: source.id.clone(),
                    path,
                    data: Map::new(),
                });
                continue;
            }
            let parsed = formats
                .deserialize(&raw, &source.format)
                .map_err(crate::persistence::format_error)?;
            let document = match parsed {
                Value::Null => Map::new(),
                Value::Object(document) => document,
                _ => {
                    return Err(EngineError::InvalidDocumentSource(Box::new(
                        InvalidDocumentSourceError {
                            source_id: source.id.clone(),
                            path: path.clone(),
                            message: format!(
                                "Document source '{}' file '{}' must contain a top-level object",
                                source.id, path
                            ),
                            collection: None,
                            id: None,
                        },
                    )))
                }
            };
            for (section_name, section_value) in &document {
                if source.collections.contains(section_name) {
                    continue;
                }
                if source.unknown_collections == UnknownCollectionPolicy::Error {
                    return Err(EngineError::UnknownCollection(Box::new(
                        UnknownCollectionError {
                            source_id: source.id.clone(),
                            path: path.clone(),
                            collection: section_name.clone(),
                            message: format!("Unknown collection '{}' in '{}'", section_name, path),
                        },
                    )));
                }
                let _ = section_value;
            }

            for collection_name in &source.collections {
                let Some(section_value) = document.get(collection_name).cloned() else {
                    continue;
                };
                let Value::Object(mut section) = section_value else {
                    return Err(EngineError::InvalidDocumentSource(Box::new(
                        InvalidDocumentSourceError {
                            source_id: source.id.clone(),
                            path: path.clone(),
                            message: format!(
                                "Collection '{}' in '{}' must be an object keyed by record id",
                                collection_name, path
                            ),
                            collection: Some(collection_name.clone()),
                            id: None,
                        },
                    )));
                };
                let file_version =
                    section.get("_version").and_then(Value::as_u64).unwrap_or(0) as u32;
                section.shift_remove("_version");
                let collection_config =
                    collection_config(config, &source.id, collection_name, Some(&path))?;
                if let Some(target_version) = collection_config.version {
                    if file_version > target_version {
                        return Err(EngineError::InvalidDocumentSource(Box::new(
                            InvalidDocumentSourceError {
                                source_id: source.id.clone(),
                                path: path.clone(),
                                message: format!(
                                    "File version {file_version} for collection '{collection_name}' is ahead of config version {target_version}"
                                ),
                                collection: Some(collection_name.clone()),
                                id: None,
                            },
                        )));
                    }
                    if file_version < target_version {
                        crate::persistence::validate_migration_registry(
                            collection_name,
                            target_version,
                            &collection_config.migrations,
                        )?;
                        if !collection_config.migrations.is_empty() {
                            section = run_migrations(
                                section,
                                file_version,
                                target_version,
                                &collection_config.migrations,
                                collection_name,
                                migration_host,
                            )?;
                        }
                    }
                }
                let decoded = decode_entity_map(
                    &path,
                    &collection_config.schema,
                    &collection_config.id_strategy,
                    section,
                    ValidationMode::Strict,
                )?;
                for (id, entity) in decoded {
                    let key = origin_key(collection_name, &id);
                    let origin = SourceRecordOrigin {
                        source_id: source.id.clone(),
                        path: path.clone(),
                        collection: collection_name.clone(),
                        id: id.clone(),
                    };
                    if let Some(first) = origins.get(&key).cloned() {
                        return Err(duplicate_record_error(collection_name, &id, first, origin));
                    }
                    let collection_entries = collections.get_mut(collection_name).ok_or_else(|| {
                        source_config_error(
                            format!(
                                "Source '{}' references collection '{}' outside the normalized collection list",
                                source.id, collection_name
                            ),
                            &source.id,
                            Some(collection_name),
                            Some(&path),
                        )
                    })?;
                    collection_entries.insert(id.clone(), entity);
                    origins.insert(key, origin);
                }
            }
            documents.push(LoadedDocument {
                source_id: source.id.clone(),
                path,
                data: document,
            });
        }
    }

    Ok(LoadedDocumentSources {
        collections,
        origins,
        documents,
    })
}

pub fn save_document_source(
    host: &dyn StorageHost,
    formats: &FormatRegistry,
    input: SaveDocumentSourceInput,
    _migration_host: Option<&dyn MigrationHost>,
) -> Result<SavedDocumentSource, EngineError> {
    let source = input
        .config
        .sources
        .iter()
        .find_map(|source| match source {
            NormalizedDatabaseSourceConfig::Documents(source) if source.id == input.source_id => {
                Some(source.clone())
            }
            _ => None,
        })
        .ok_or_else(|| {
            EngineError::InvalidDocumentSource(Box::new(InvalidDocumentSourceError {
                source_id: input.source_id.clone(),
                path: String::new(),
                message: format!("Unknown document source '{}'", input.source_id),
                collection: None,
                id: None,
            }))
        })?;

    let mut projected_by_path = IndexMap::new();
    for document in input
        .documents
        .iter()
        .filter(|document| document.source_id == source.id)
    {
        projected_by_path.insert(
            document.path.clone(),
            clone_document_for_source(document, &source, &input.config),
        );
    }

    let mut new_origins = input.origins.clone();
    for (key, origin) in input.origins.iter() {
        if origin.source_id != source.id || !source.collections.contains(&origin.collection) {
            continue;
        }
        let exists = input
            .collections
            .get(&origin.collection)
            .map(|collection| collection.contains_key(&origin.id))
            .unwrap_or(false);
        if !exists {
            new_origins.shift_remove(key);
            if let Some(document) = projected_by_path.get_mut(&origin.path) {
                if let Some(Value::Object(section)) = document.get_mut(&origin.collection) {
                    section.shift_remove(&origin.id);
                }
            }
        }
    }

    for collection_name in &source.collections {
        let Some(collection) = input.collections.get(collection_name) else {
            continue;
        };
        let collection_config = collection_config(
            &input.config,
            &source.id,
            collection_name,
            Some(&source.root),
        )?;
        for (id, entity) in collection {
            let key = origin_key(collection_name, id);
            let existing_origin = input.origins.get(&key);
            if let Some(existing_origin) = existing_origin {
                if existing_origin.source_id != source.id {
                    continue;
                }
            }
            let path = existing_origin
                .map(|origin| origin.path.clone())
                .unwrap_or_else(|| source.outbox.clone());
            let document = projected_by_path
                .entry(path.clone())
                .or_insert_with(Map::new);
            let section_value = document
                .entry(collection_name.clone())
                .or_insert_with(|| Value::Object(Map::new()));
            let section = section_value.as_object_mut().ok_or_else(|| {
                EngineError::InvalidDocumentSource(Box::new(InvalidDocumentSourceError {
                    source_id: source.id.clone(),
                    path: path.clone(),
                    message: format!(
                        "Collection '{}' in '{}' must be an object keyed by record id",
                        collection_name, path
                    ),
                    collection: Some(collection_name.clone()),
                    id: None,
                }))
            })?;
            if let Some(version) = collection_config.version {
                section.insert("_version".to_owned(), Value::Number(version.into()));
            }
            let encoded = encode_value(
                &collection_config.schema,
                &strip_derived_id_field(entity, &collection_config.id_strategy),
                id,
            )?;
            section.insert(id.clone(), encoded);
            new_origins.insert(
                key,
                SourceRecordOrigin {
                    source_id: source.id.clone(),
                    path: path.clone(),
                    collection: collection_name.clone(),
                    id: id.clone(),
                },
            );
        }
    }

    let mut merged_documents = input
        .documents
        .into_iter()
        .filter(|document| document.source_id != source.id)
        .collect::<Vec<_>>();
    let writes = projected_by_path
        .iter()
        .map(|(path, data)| {
            formats
                .serialize(&Value::Object(data.clone()), &source.format, None)
                .map(|raw| (path.clone(), raw, data.clone()))
                .map_err(crate::persistence::format_error)
        })
        .collect::<Result<Vec<_>, _>>()?;
    for (path, raw, data) in writes {
        host.ensure_dir(&path)?;
        host.write(&path, &raw)?;
        merged_documents.push(LoadedDocument {
            source_id: source.id.clone(),
            path,
            data,
        });
    }
    merged_documents.sort_by(|left, right| left.path.cmp(&right.path));

    Ok(SavedDocumentSource {
        origins: new_origins,
        documents: merged_documents,
    })
}
